ex5/FileSystemTPL/TPLOperations/MoveTPL.cs

```diff
@@ -0,0 +1,109 @@
+﻿using System;
+using System.IO;
+using System.Linq;
+
+namespace FileSystemTPL.TPLOperations
+{
+    class MoveTPL : Operation
+    {
+        public MoveTPL(CommandDetails commandDetails, ExecutionManager executionManager) :
+            base(commandDetails, executionManager)
+        { }
+
+        public override void Execute()
+        {
+            string source = Path.Combine(Program.DIRECTORY_PATH, commandDetails.CommandArgs.First());
+            string destination = Path.Combine(Program.DIRECTORY_PATH, commandDetails.CommandArgs.Last());
+
+            Action<string, string> MovingFunction = GetMovingFunction(source);
+            MovingFunction(source, destination);
+        }
+
+        private Action<string, string> GetMovingFunction(string source)
+        {
+            try
+            {
+                executionManager.LockPath(source, ObjectInActionStatus.ToBeMoved);
+                
+                if (File.Exists(source))
+                {
+                    return MoveFileTPL;
+                }
+
+                else if (Directory.Exists(source))
+                {
+                    return MoveDirRecursively;
+                }
+
+                else
+                {
+                    throw new FileNotFoundException($"The path {source} does not exist.");
+                }
+            }
+
+            finally
+            {
+                executionManager.ReleasePath(source);
+            }
+        }
+
+        private void MoveFileTPL(string source, string destination)
+        {
+            try
+            {
+                executionManager.LockPath(destination, ObjectInActionStatus.MovingDestination);
+                basicOperations.MoveFile(source, destination);
+            }
+
+            finally
+            {
+                executionManager.ReleasePath(destination);
+            }
+        }
+
+        private void MoveDirRecursively(string source, string destination)
+        {
+            try
+            {
+                executionManager.LockPath(destination, ObjectInActionStatus.MovingDestination);
+                CreateDestinationDir(destination);
+                MoveChildFiles(source, destination);
+                MoveChildDirs(source, destination);
+                DeleteSourceDir(source);
```

Comment:
What happens when the source contained in the target (or the opposite) or when source=target?

Reply:
If the source is a subfolder of the target there is no problem, there is one in the opposite case and in case where they are equal. I will add a check for these cases.